### PR TITLE
fix: mscolab_chat, timeout tuple problem

### PR DESCRIPTION
**Pull request by levi178u**
fixes: #2716

What are the changes ?

- Updated default config to provide timeout as a tuple instead of a list.
- Modified validation logic to expect tuples for timeout.
- Enhanced merge_dict() to convert user-provided lists to tuples during merge.
- Ensured compatibility with requests.get() by enforcing correct format at config level.

### CODE REVIEW

**Comment by levi178u** [comment]
Mentors plz review the PR

**Review by ReimarBauer — CHANGES_REQUESTED** [review]
please look on what you pushed.

which changes are related to the issue?

Could you also add a unit test which shows the problem happening and that your suggestion does fix it.

**Comment by levi178u** [comment]
> please look on what you pushed.
> 
> which changes are related to the issue?
> 
> Could you also add a unit test which shows the problem happening and that your suggestion does fix it.

Basically the `requests.get()` funct is getting `([2, 10], [2, 10])` but it expects it to be `([2,10]) ` i.e `(connect_timeout, read_timeout)`. So in the `config.py` I just made sure the above is done and I tested it via `_test_utils `'s `test_config` pytest , where all cases passed and `pytest.log` was auto-generated.

**Comment by levi178u** [comment]
I also wanna know what other changes should I do so that this timeout config. issue is resolved.

**Comment by ReimarBauer** [comment]
the idea for getting knowledge if it is complete is to write first a test which shows the problem. Then provide the fix, which does not anymore have a failing test

**Comment by ReimarBauer** [comment]
look on the changes url yourselft. What is irrelevant for the issue?

Bildschirmfoto 2025-04-10 um 10 49 45

**Comment by levi178u** [comment]
> look on the changes url yourselft. What is irrelevant for the issue?
> 
> Bildschirmfoto 2025-04-10 um 10 49 45

Yes in the changes section the `"MSCOLAB_timeout": ([2, 10])` and     `MSCOLAB_timeout = (2, 10)` is only relevent as I'm solving the tuple issue and i had done changes in the `config.py` under `utils`, also i added `fs` in `pixi.toml` coz I ran the pytest for this using this: 
```
 pixi install
pixi run -e dev pytest tests/_test_utils
```

image
Here's the test log, once review this @ReimarBauer

**Comment by levi178u** [comment]
> I also wanna know what other changes should I do so that this timeout config. issue is resolved.

Ok I got it, I'll write some test cases against the changes,sure.

File: pixi.toml [hunk 1]
```diff
@@ -26,6 +26,7 @@ flask-socketio = ">=5.1.0"
 flask-sqlalchemy = ">=3.0.0"
 flask-wtf = "*"
 fs_filepicker = "*"
+fs = "*"
```
Comment (ReimarBauer) [inline pixi.toml]: this is inherited by fs_filepicker, what is the reason to add it?

File: pytest.log [hunk 2]
```diff
@@ -0,0 +1,141 @@
+2025-04-09 19:09:01 DEBUG Failed to pyproj from mpl_toolkits.basemap
```
Comment (ReimarBauer) [inline pytest.log]: this file is not wanted to be committed

File: pixi.toml (hunk 1 above)
Comment (levi178u) [reply pixi.toml]: Yes your right, I'll remove it, since I just added it so as to check the pytest suite as the file had `import fs`, no need for it to be committed as such

File: pytest.log (hunk 2 above)
Comment (levi178u) [reply pytest.log]: Yes i'll make changes

**Review by ReimarBauer — CHANGES_REQUESTED** [review]
The test case does not accurately reflect the scenario related to this issue. The proposed fix breaks the config editor. Test module names must start with "test_".

With only 70% test coverage, we need to exercise caution when modifying code outside of known issue areas. The best practice is to create a failing test case that demonstrates the problem before implementing a fix.  This approach will also contribute to increasing our overall test coverage.

https://github.com/user-attachments/assets/18c68c3a-be33-44a9-8ebb-b369c09d5324

**Comment by levi178u** [comment]
Ok @ReimarBauer  I have prepared a test case file covering more cases and edge cases too, can I push it for review. Also I have asked few doubts in slack thread, do check them. Cases such as negative nums, other ds besides tuple, empty ds, float nums, int nums, etc.. i have added cases considering them too

**Comment by ReimarBauer** [comment]
> #2716

The issue is not assigned to you. But when you want to work on it reproduce first by a test the issue description.

**Comment by levi178u** [comment]
> > #2716
> 
> The issue is not assigned to you. But when you want to work on it reproduce first by a test the issue description.

Ok , here I have added the test file (`test_mscolab_timeout.py`) , pls review the same, also what changes should I do in the test file pls suggest

**Review by ReimarBauer — COMMENTED** [review]
A solution has to succeed this test:

https://github.com/Open-MSS/MSS/blob/stable/tests/_test_msui/test_mscolab.py#L301

**Comment by ReimarBauer** [comment]
solved by https://github.com/Open-MSS/MSS/pull/2815
